### ISSUE TITLE
Promise chains fork with MutableData leading to race conditions

### ISSUE THREAD

**Issue by btbonval** [loziniak/maidsafe-test#1]
The following code forks the Promise into two separate then threads: https://github.com/loziniak/maidsafe-test/blob/630d5fc6ec7cdd1ab04dba276f580f88d9a83641/safe_api.js#L37

If the `App.handlePromise` chain hits `App.free()` before `MutableData.then()` completes (which always appears to be the case in my mocknet tests), then the `MutableData.handlePromise` chain is operating on a freed appHandle. I'm pretty sure this has been the source of many problems I've had trying to get the code to work.

I have written some work around code (will be submitted as a PR) so that the following works as expected:
```
        let app = safeAPI.initialiseApp(app_info);
        app.authoriseAndConnect({_public: ['Read']});
        let md = app.quickRandomPublic({'this': 'is', 'a': 'test'});
        md.getString('this').then(console.log);
        md.getString('a').then(console.log);
        app.then(_ => console.log('All done!'));
        md.free();
        app.catch(console.log);
        app.free();
```

Above code is based on Example 2 in https://github.com/maidsafe/safe_examples/releases/download/0.13.0/examples.html

**Comment by loziniak** [comment]
The more I think about the issue, the more I'm convinced that forking the promises is a bad thing, and I have no idea how to solve the issue in an elegant way.

**Comment by loziniak** [comment]
fixed in #2 by @btbonval
